compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3460,32 +3460,63 @@ TR::Register *OMR::Power::TreeEvaluator::vmulEvaluator(TR::Node *node, TR::CodeG
    {
    switch(node->getDataType())
      {
+     case TR::VectorInt8:
+       return TR::TreeEvaluator::vmulInt8Helper(node, cg);
+     case TR::VectorInt16:
+       return TR::TreeEvaluator::vmulInt16Helper(node,cg);
      case TR::VectorInt32:
        return TR::TreeEvaluator::vmulInt32Helper(node,cg);
      case TR::VectorFloat:
        return TR::TreeEvaluator::vmulFloatHelper(node,cg);
      case TR::VectorDouble:
        return TR::TreeEvaluator::vmulDoubleHelper(node,cg);
-     case TR::VectorInt16:
-       return TR::TreeEvaluator::vmulInt16Helper(node,cg);
      default:
        TR_ASSERT(false, "unrecognized vector type %s\n", node->getDataType().toString()); return NULL;
      }
    }
 
-TR::Register *OMR::Power::TreeEvaluator::vmulInt32Helper(TR::Node *node, TR::CodeGenerator *cg)
+TR::Register *OMR::Power::TreeEvaluator::vmulInt8Helper(TR::Node *node, TR::CodeGenerator *cg)
    {
-   return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmuluwm);
-   }
+   TR::Node *firstChild;
+   TR::Node *secondChild;
+   TR::Register *lhsReg, *rhsReg;
+   TR::Register *productReg;
+   TR::Register *temp;
+   TR::Register *shiftReg;
 
-TR::Register *OMR::Power::TreeEvaluator::vmulFloatHelper(TR::Node *node, TR::CodeGenerator *cg)
-   {
-   return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::xvmulsp);
-   }
+   firstChild = node->getFirstChild();
+   secondChild = node->getSecondChild();
 
-TR::Register *OMR::Power::TreeEvaluator::vmulDoubleHelper(TR::Node *node, TR::CodeGenerator *cg)
-   {
-   return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::xvmuldp);
+   lhsReg = cg->evaluate(firstChild);
+   rhsReg = cg->evaluate(secondChild);
+
+   productReg = cg->allocateRegister(TR_VRF);
+   temp = cg->allocateRegister(TR_VRF);
+   shiftReg = cg->allocateRegister(TR_VRF);
+
+   node->setRegister(productReg);
+
+   //set shift amount to 1 byte
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisb, node, shiftReg, 8);
+
+   //multiply even bytes and shift to correct location
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vmuleub, node, productReg, lhsReg, rhsReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vslh, node, productReg, productReg, shiftReg);
+
+   //multiply odd byte and shift right (to discard upper byte of product) and then left (to relocate to correct position)
```

Comment:
This comment says that we shift right to discard the upper byte, and then left to put the remaining bits back, but that sequence of shifts would discard the *lower* byte instead. This is just a problem with the comment though—the code correctly shifts first left and then right